src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,117 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (ShouldRestoreNullValues())
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties;
+
+                        if (resourceState.SelectedProperties == SelectedPropertiesNode.EntireSubtree)
+                        {
+                            selectedProperties = GetSelfAndBaseTypeProperties(edmStructuredType);
+                        }
+                        else
+                        {
+                           // Partial subtree.
+                           selectedProperties = resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType);
+                        }
+
+                        // All dynamic properties: null value restoration.
+                        foreach (string name in resourceState.SelectedProperties.GetSelectedDynamicProperties(edmStructuredType))
+                        {
+                            RestoreNullODataProperty(name, resourceState);
+                        }

```

Comment:
What if the dynamic property already existed in the payload; this seems like it would add the property a second time with the null value.

Reply:
Resolved. Thanks.

---
In reply to: [198063535](https://github.com/OData/odata.net/pull/1106#discussion_r198063535) [](ancestors = 198063535)